javascript課題3-2　反映

rui-ishikawa opened the pull request:
javascript課題3-2

[変更点]
・削除機能を追加
削除ボタンの挙動をonClick属性で実装しています。
addEventListenerを使ってみたのですが、上手くイベントをハンドリングすることができませんでした。（複数のTODOを追加しても、一番上の要素に対してしか削除処理が発火しませんでした）
可能であれば、addEventListenerを使った実装方法をご教示いただきたいです。

rui-ishikawa committed:
[add] javascript課題3-2 削除機能追加

rui-ishikawa committed:
[modify] 入力フォームクリア処理を追加、メソッド名修正、ダブルクォーテーションに統一

rui-ishikawa committed:
[modify] javascript3-2反映

rui-ishikawa committed:
[modify] ソースファイルの読み込み修正

rui-ishikawa committed:
[modify] 不要ファイルの削除

kouki-iwahara commented on javascript/task3-1/task3_1.js:
```diff
@@ -1,14 +1,28 @@
+  createNode();
+};
+
+function createNode() {
```
ボリュームが出てきましたので、更に処理を関数に切り分けましょう。
・状態ボタンを生成
・削除ボタンを生成
あたりをとりあえず切り分ければ見通しが良くなるかなと思います。

rui-ishikawa replied:
状態ボタン作成と削除ボタンの作成を別だししました。

kouki-iwahara commented on javascript/task3-1/task3_1.js:
```diff
@@ -35,6 +49,7 @@ function displayTodos() {
     const tdDeleteInput = document.createElement("input");
     tdDeleteInput.setAttribute("type","button");
 		tdDeleteInput.setAttribute("value","削除");
+		tdDeleteInput.setAttribute("onClick","reDisplayTodos()");
```
onClickではなく、addEventListenerを使用しましょう。
onClickは既に同名の属性があった場合には上書されてしまいます。

reDisplayTodosですが、DisplayTodosという名前の関数も使用していることから名前が紛らわしいので、素直にdeleteTodoを発火させれば良いかなと思います。

kouki-iwahara commented on javascript/task3-1/task3_1.js:
```diff
@@ -44,15 +59,20 @@ function displayTodos() {
 
-let todos = [];
+function deleteTodo() {
+  const itemId = event.target.parentNode.parentNode.id;
```
クリックされたtodoを引数に取るようにすると、ここの記述がもっとスマートになるかなと思います。

rui-ishikawa replied:
ご指摘の通り修正しました。

kouki-iwahara commented:
>addEventListenerを使ってみたのですが、上手くイベントをハンドリングすることができませんでした。（複数のTODOを追加しても、一番上の要素に対してしか削除処理が発火しませんでした）

コメントにも記述しましたが、addEventListenerを使用した方が良いですね。
現在は配列todosにtodoが格納されていると思います。
todosの中の何番目のtodoがクリックされたかがわかればそのtodoを操作できますね^^
以下参考記事です。
https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf

rui-ishikawa committed:
[modify]
・削除イベントの」ハンドリングにaddEventLisnerを使うよう修正
・statusボタン、削除ボタンの作成を別だし
・createNode から initializeNodeにメソッド名を変更

rui-ishikawa commented:
何番目のtodoがクリックされたかという情報を取得するうまい方法がわかりません。
ノード初期化時に、querySelectorAllで要素を取得して、addEventListenerでクリックイベントを登録しています。その時に、インデックスの取得と削除処理を行うよう実装しています。
このあたりベストプラクティスがあれば教えていただきたいです。

kouki-iwahara commented:
>ノード初期化時に、querySelectorAllで要素を取得して、addEventListenerでクリックイベントを登録しています。

大分試行錯誤したのですね^^
かなり頑張ったのが伝わってきます！
少し解説します。

まず、削除ボタンを生成した時に、一緒にイベントの付与まですると良いかなと思います。
そのイベントで発火される関数に引数を持たせ、その引数には、新たに作成されたtodoかindexが入るようにすればもっとすっきり書けますね！
下記のイメージです。
```
function deleteTodo(index) {
  todos.splice(index, 1);
}

function createDeleteButton(todo) {
  const deleteBtn = document.getElementById('btn');
  const index = todos.indexOf(todo);
  deleteBtn.addEventListener('click', () => {
    deleteTodo(index);
  });
  return deleteBtn
}
```

まず、追加ボタンが押された時にdisplayTodosではなくaddTodoを発火させ、addTodoの中でdisplayTodosを発火させれば引数にtodoが持てますね。
そしてdisplayTodosの中でcreateDeleteButtonを発火させ、引数にtodoを持たせて...という感じで書くと、大分わかりやすくなるかなと思います。(registerDeleteEventが不要になる)

あと、initializeNodeという名前がややこしくさせてしまっている気がするので、あえて切り分けなくてもdisplayTodosに記述してしまって良いかなと思います。
処理の内容も初期化ではなくtodoを表示させていますので。

rui-ishikawa committed:
[modify] createDeleteButton内でイベント登録処理を実装、追加ボタン押下時にaddTodoが発火するよう修正

engineer-ruirui commented:
【変更点】
・削除ボタン生成時にイベント登録を実施
・引数にindexを持たせるよう修正
・追加ボタン押下時にaddTodoが発火するよう修正
前よりは良いですが、上記のように上手く実装できないです・・・

【質問】
削除ボタン作成処理の中で、getElementByIdした要素をリターンしていますが、
・この関数の呼び方
・リターンした先での処理（appendChildなどしていると思っているのですが）
がわかりませんでした。
こちら教えていただけますでしょうか。

kouki-iwahara commented on javascript/task3-1/task3_1.js:
```diff
@@ -1,58 +1,74 @@
-  addTodo()
-
   // 配列の要素分子ノードを追加
   for (let i = 0; i < todos.length; i++) {
```
配列の要素を回すときは、今回はforEachを使用すると良いです！
https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach

配列の値操作は便利なものがたくさんありますので、どんどん使用していきましょう！

kouki-iwahara commented on javascript/task3-1/task3_1.js:
```diff
@@ -1,58 +1,74 @@
-    tdDelete.appendChild(tdDeleteInput);
-    tr.appendChild(tdDelete);
-
+    tr.appendChild(createDeleteButton(i));
```
const hoge = createDeleteButton(i)
のように一度変数に格納すると良いでしょう。
扱いやすくもなりますね！

kouki-iwahara commented:
>削除ボタン作成処理の中で、getElementByIdした要素をリターンしていますが、
・この関数の呼び方
・リターンした先での処理（appendChildなどしていると思っているのですが）
をどのように実装したら良いかわかりませんでした。

コメントにも書きましたが、一度変数に格納する形で記述した方が良いですね。
あとは問題ないと思います！

rui-ishikawa committed:
[modify] forEachを使うように修正、ボタン作成処理を変数に格納するよう修正

engineer-ruirui commented:
ありがとうございます！
ご指摘の通り修正しました！

manaberuit reviewed (approved):
OKです！
